From a49d454a019334f05dc7f6ba9a96299a178ddd19 Mon Sep 17 00:00:00 2001
From: Tim Lunn <tim@feathertop.org>
Date: Sun, 1 Sep 2013 09:34:17 +1000
Subject: [PATCH] Revert "Revert "media-keys: Add key bindings to switch input
 sources""

This reverts commit cc45fec342713745f391533a202976e97670f9e5.
Patched to use the new g_settings keys for input source switching
---
 plugins/media-keys/gsd-media-keys-manager.c | 59 ++++++++++++++++++++++++++++-
 plugins/media-keys/shortcuts-list.h         |  6 +++
 2 files changed, 64 insertions(+), 1 deletion(-)

Index: gnome-settings-daemon-3.8.5/plugins/media-keys/gsd-media-keys-manager.c
===================================================================
--- gnome-settings-daemon-3.8.5.orig/plugins/media-keys/gsd-media-keys-manager.c	2013-11-12 16:32:40.294469502 +1300
+++ gnome-settings-daemon-3.8.5/plugins/media-keys/gsd-media-keys-manager.c	2013-11-12 16:32:40.286469502 +1300
@@ -114,6 +114,10 @@
 #define VOLUME_STEP 6           /* percents for one volume button press */
 #define MAX_VOLUME 65536.0
 
+#define GNOME_DESKTOP_INPUT_SOURCES_DIR "org.gnome.desktop.input-sources"
+#define KEY_CURRENT_INPUT_SOURCE "current"
+#define KEY_INPUT_SOURCES        "sources"
+
 #define SYSTEMD_DBUS_NAME                       "org.freedesktop.login1"
 #define SYSTEMD_DBUS_PATH                       "/org/freedesktop/login1"
 #define SYSTEMD_DBUS_INTERFACE                  "org.freedesktop.login1.Manager"
@@ -157,6 +161,7 @@
 #endif /* HAVE_GUDEV */
 
         GSettings       *settings;
+        GSettings       *input_settings;
         GHashTable      *custom_settings;
 
         GPtrArray       *keys;
@@ -530,7 +535,9 @@
 get_key_string (GsdMediaKeysManager *manager,
 		MediaKey            *key)
 {
-	if (key->settings_key != NULL)
+        if (key->settings_key == "switch-input-source" || key->settings_key == "switch-input-source-backward")
+	        return g_settings_get_strv (manager->priv->input_settings, key->settings_key)[0];
+        else if (key->settings_key != NULL)
 		return g_settings_get_string (manager->priv->settings, key->settings_key);
 	else if (key->hard_coded != NULL)
 		return g_strdup (key->hard_coded);
@@ -2111,6 +2118,44 @@
                  * are not used in this context */
                 break;
         }
+
+}
+
+static void
+do_switch_input_source_action (GsdMediaKeysManager *manager,
+                               MediaKeyType         type)
+{
+        GSettings *settings;
+        GVariant *sources;
+        gint i, n;
+
+        if (!manager->priv->have_legacy_keygrabber)
+                return;
+
+        settings = g_settings_new (GNOME_DESKTOP_INPUT_SOURCES_DIR);
+        sources = g_settings_get_value (settings, KEY_INPUT_SOURCES);
+
+        n = g_variant_n_children (sources);
+        if (n < 2)
+                goto out;
+
+        i = g_settings_get_uint (settings, KEY_CURRENT_INPUT_SOURCE);
+
+        if (type == SWITCH_INPUT_SOURCE_KEY)
+                i += 1;
+        else
+                i -= 1;
+
+        if (i < 0)
+                i = n - 1;
+        else if (i >= n)
+                i = 0;
+
+        g_settings_set_uint (settings, KEY_CURRENT_INPUT_SOURCE, i);
+
+ out:
+        g_variant_unref (sources);
+        g_object_unref (settings);
 }
 
 static void
@@ -2451,6 +2496,10 @@
         case BATTERY_KEY:
                 do_battery_action (manager);
                 break;
+        case SWITCH_INPUT_SOURCE_KEY:
+        case SWITCH_INPUT_SOURCE_BACKWARD_KEY:
+                do_switch_input_source_action (manager, type);
+                break;
         /* Note, no default so compiler catches missing keys */
         case CUSTOM_KEY:
                 g_assert_not_reached ();
@@ -2773,6 +2822,12 @@
         g_signal_connect (G_OBJECT (manager->priv->settings), "changed::custom-keybindings",
                           G_CALLBACK (gsettings_custom_changed_cb), manager);
 
+        manager->priv->input_settings = g_settings_new (INPUT_SETTINGS_BINDING_DIR);
+        g_signal_connect (G_OBJECT (manager->priv->input_settings), "changed",
+                          G_CALLBACK (gsettings_changed_cb), manager);
+        g_signal_connect (G_OBJECT (manager->priv->input_settings), "changed::custom-keybindings",
+                          G_CALLBACK (gsettings_custom_changed_cb), manager);
+
         manager->priv->custom_settings =
           g_hash_table_new_full (g_str_hash, g_str_equal,
                                  g_free, g_object_unref);
@@ -2903,6 +2958,7 @@
 
         g_clear_object (&priv->logind_proxy);
         g_clear_object (&priv->settings);
+        g_clear_object (&priv->input_settings);
         g_clear_object (&priv->power_settings);
         g_clear_object (&priv->power_proxy);
         g_clear_object (&priv->power_screen_proxy);
Index: gnome-settings-daemon-3.8.5/plugins/media-keys/shortcuts-list.h
===================================================================
--- gnome-settings-daemon-3.8.5.orig/plugins/media-keys/shortcuts-list.h	2013-11-12 16:32:40.294469502 +1300
+++ gnome-settings-daemon-3.8.5/plugins/media-keys/shortcuts-list.h	2013-11-12 16:32:40.286469502 +1300
@@ -25,6 +25,7 @@
 #include "gsd-keygrab.h"
 
 #define SETTINGS_BINDING_DIR "org.gnome.settings-daemon.plugins.media-keys"
+#define INPUT_SETTINGS_BINDING_DIR "org.gnome.desktop.wm.keybindings"
 
 typedef enum {
         TOUCHPAD_KEY,
@@ -83,6 +84,8 @@
         KEYBOARD_BRIGHTNESS_DOWN_KEY,
         KEYBOARD_BRIGHTNESS_TOGGLE_KEY,
         BATTERY_KEY,
+        SWITCH_INPUT_SOURCE_KEY,
+        SWITCH_INPUT_SOURCE_BACKWARD_KEY,
         CUSTOM_KEY
 } MediaKeyType;
 
@@ -160,6 +163,9 @@
         { KEYBOARD_BRIGHTNESS_DOWN_KEY, NULL, N_("Keyboard Brightness Down"), "XF86KbdBrightnessDown", SHELL_KEYBINDING_MODE_ALL },
         { KEYBOARD_BRIGHTNESS_TOGGLE_KEY, NULL, N_("Keyboard Brightness Toggle"), "XF86KbdLightOnOff", SHELL_KEYBINDING_MODE_ALL },
         { BATTERY_KEY, NULL, N_("Battery Status"), "XF86Battery", GSD_KEYBINDING_MODE_LAUNCHER },
+        { SWITCH_INPUT_SOURCE_KEY, "switch-input-source", NULL, NULL, SHELL_KEYBINDING_MODE_ALL },
+        { SWITCH_INPUT_SOURCE_BACKWARD_KEY, "switch-input-source-backward", NULL, NULL, SHELL_KEYBINDING_MODE_ALL }
+
 };
 
 #undef SCREENSAVER_MODE
